tedge/src/cli/connect/command.rs

```diff
@@ -278,52 +292,69 @@ async fn create_device(port: u16) -> Result<DeviceStatus, ConnectError> {
+fn check_device_status_azure(port: u16) -> Result<DeviceStatus, ConnectError> {
     const AZURE_TOPIC_DEVICE_TWIN_DOWNSTREAM: &str = r##"az/twin/res/#"##;
     const AZURE_TOPIC_DEVICE_TWIN_UPSTREAM: &str = r#"az/twin/GET/?$rid=1"#;
-    const CLIENT_ID: &str = "check_connection_az";
-
-    let device_twin_pub_topic = Topic::new(AZURE_TOPIC_DEVICE_TWIN_UPSTREAM)?;
-    let device_twin_sub_filter = TopicFilter::new(AZURE_TOPIC_DEVICE_TWIN_DOWNSTREAM)?;
-
-    let mqtt = Client::connect(CLIENT_ID, &mqtt_client::Config::default().with_port(port)).await?;
-    let mut device_twin_response = mqtt.subscribe(device_twin_sub_filter).await?;
-
-    let (sender, mut receiver) = tokio::sync::oneshot::channel();
-
-    let _task_handle = tokio::spawn(async move {
-        if let Some(message) = device_twin_response.next().await {
-            //status should be 200 for successful connection
-            if message.topic.name.contains("200") {
-                let _ = sender.send(true);
-            } else {
-                let _ = sender.send(false);
+    const CLIENT_ID: &str = "c\
+    heck_connection_az";
+    const REGISTRATION_PAYLOAD: &[u8] = b"";
+    const REGISTRATION_OK: &str = "200";
+
+    let mut options = MqttOptions::new(CLIENT_ID, DEFAULT_HOST, port);
+    options.set_keep_alive(RESPONSE_TIMEOUT.as_secs() as u16);
+
+    let (mut client, mut connection) = rumqttc::Client::new(options, 10);
+    let mut acknowledged = false;
+
+    client.subscribe(AZURE_TOPIC_DEVICE_TWIN_DOWNSTREAM, AtLeastOnce)?;
+
+    for event in connection.iter() {
+        match event {
+            Ok(Event::Incoming(Packet::SubAck(_))) => {
+                // We are ready to get the response, hence send the request
+                client.publish(
+                    AZURE_TOPIC_DEVICE_TWIN_UPSTREAM,
+                    AtLeastOnce,
+                    false,
+                    REGISTRATION_PAYLOAD,
+                )?;
+            }
+            Ok(Event::Incoming(Packet::PubAck(_))) => {
+                // The request has been sent
+                acknowledged = true;
+            }
+            Ok(Event::Incoming(Packet::Publish(response))) => {
+                // We got a response
+                if response.topic.contains(REGISTRATION_OK) {
+                    return Ok(DeviceStatus::AlreadyExists);
+                } else {
+                    break;
+                }
             }
+            Ok(Event::Outgoing(Outgoing::PingReq)) => {
+                // No messages have been received for a while
+                println!("\nLocal MQTT publish has timed out.");
+                break;
+            }
+            Ok(Event::Incoming(Incoming::Disconnect)) => {
+                eprintln!("ERROR: Disconnected");
+                break;
+            }
+            Err(err) => {
+                eprintln!("ERROR: {:?}", err);
+                break;
+            }
+            _ => {}
         }
```

Comment:
Correct me if I'm wrong but I believe this will wait for a message until either it receives one or until mqtt ping will fail (connection timeout), wouldn't we want to have a little bit more control over the timeout and how long it takes to wait for the test to finish?

Reply:
You are correct. The ping period has been set to the target timeout though.

```options.set_keep_alive(RESPONSE_TIMEOUT.as_secs() as u16);```

Reply:
Yes, this makes sense, I missed that line, thanks.